app/apollo/resolvers/group.js

```diff
@@ -559,21 +638,55 @@ const groupResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({ req_id, user, org_id, uuid, clusters }, `${queryName} validating`);
+        logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} validating`);
 
         validateString( 'org_id', org_id );
         validateString( 'uuid', uuid );
         clusters.forEach( value => validateString( 'clusters', value ) );
 
-        // validate the group exits in the db first.
         const group = await models.Group.findOne({ org_id: org_id, uuid });
-        if(!group){
+        logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} validating - found: ${!!group}`);
+
+        const identifiers = group ? [uuid, group.name] : [uuid];
+        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, identifiers);
+        logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} validating - authorized`);
+
+        if (!group) {
           throw new NotFoundError(context.req.t('group uuid "{{uuid}}" not found', {'uuid':uuid}), context);
         }
 
-        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, [group.uuid, group.name]);
+        let allAllowed = false;
+        try {
+          await validAuth(me, org_id, ACTIONS.READ, TYPES.CLUSTER, queryName, context);
+          allAllowed = true;
+        }
+        catch(e){ // If exception thrown, user does NOT have auth to all resources of this type, and code must later filter based on fine grained auth
+        }
+        logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} validating - allAllowed: ${allAllowed}`);
+
+        let foundClusters = await commonClusterSearch(models, {org_id}, { limit: 0, skip: 0, startingAfter: null });
+        foundClusters = foundClusters.filter(cluster => clusters.includes(cluster.cluster_id));
+
+        if (!allAllowed) {
+          foundClusters = await filterResourcesToAllowed(me, org_id, ACTIONS.READ, TYPES.CLUSTER, foundClusters, context);
+          logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} filtered resources to allowed`);
+        }
+
+        if(!clusters.length){
+          throw new ValidationError(context.req.t('No clusters were passed', {'clusters':clusters}), context);
+        }
```

Comment:
If you are going to add a check for whether the input `clusters` is not empty:
- Use `if( clusters.length < 1 )` rather than `if(!clusters.length)` -- more readable at the very least, we're trying to see if the length is zero, not trying to see if the length is 'false-y'
- This is a 'simple' check -- there's no reason to wait and check this only after the expensive database and authorization checks.  If included, it should be after `validateString` statements and before `clusters.forEach` validation around line 645.

Reply:
fixed

Reply:
Resolved